src/extension/mapml/src/test/java/org/geoserver/mapml/tcrs/CustomTiledCRSTest.java

```diff
@@ -0,0 +1,229 @@
+    }
+
+    protected void onSetUp(SystemTestData testData) throws Exception {
+        super.onSetUp(testData);
+
+        GeoServer gs = getGeoServer();
+        GeoServerInfo global = gs.getGlobal();
+        Catalog catalog = gs.getCatalog();
+
+        // add UTM31 Gridset to GWC
+        GridSetBroker broker = GWC.get().getGridSetBroker();
+        broker.addGridSet(createUtm31NGridset());
+
+        // Add 2 custom gridsets to TCRS list
+        MetadataMap metadata = global.getSettings().getMetadata();
+        ArrayList<String> crsList = new ArrayList<>();
+        crsList.add("UTM31WGS84Quad");
+        metadata.put(TiledCRSConstants.TCRS_METADATA_KEY, crsList);
+        gs.save(global);
+
+        ResourceInfo layerMeta =
+                catalog.getLayerByName(MockData.ROAD_SEGMENTS.getLocalPart()).getResource();
+        layerMeta.getMetadata().put("mapml.useTiles", false);
+        catalog.save(layerMeta);
+        TiledCRSConstants.reloadDefinitions();
+    }
+
+    private GridSet createUtm31NGridset() {
+        SRS srs = SRS.getSRS(32631);
+        BoundingBox bbox = new BoundingBox(166021.44, 0.0, 833978.56, 9329005.18); // In meters
+        double[] resolutions =
+                new double[] {
+                    156543.03392804097, 78271.51696402048, 39135.75848201024, 19567.87924100512,
+                    9783.93962050256, 4891.96981025128, 2445.98490512564, 1222.99245256282,
+                    611.49622628141, 305.748113140705, 152.8740565703525, 76.43702828517625,
+                    38.21851414258813, 19.109257071294063, 9.554628535647032, 4.777314267823516
+                };
+
+        double[] scaleDenoms = null;
+        Double metersPerUnit = 1.0;
+        double pixelSize = 0.00028;
+        String[] scaleNames = null;
+
+        int tileWidth = 256;
+        int tileHeight = 256;
+        boolean yCoordinateFirst = false;
+        GridSet gridSet =
+                GridSetFactory.createGridSet(
+                        "UTM31WGS84Quad", // GridSet name
+                        srs, // Spatial reference system (EPSG:32631)
+                        bbox, // Bounding box
+                        true, // Align top-left
+                        resolutions, // Resolutions (meters per pixel)
+                        scaleDenoms, // Scale denominators (can be null)
+                        metersPerUnit, // Meters per unit
+                        pixelSize, // Pixel size in meters (0.00028 for WMS)
+                        scaleNames, // Scale names (optional, can be null)
+                        tileWidth, // Tile width (256 pixels)
+                        tileHeight, // Tile height (256 pixels)
+                        yCoordinateFirst // Whether Y-coordinate comes first (false)
+                        );
+
+        return gridSet;
+    }
+
+    @Test
+    public void testGridsetToTiledCRS() throws Exception {
+        CoordinateReferenceSystem crs = CRS.decode("MapML:UTM31WGS84Quad");
```

Comment:
Good. I would also check that:
* The code shows up in "CRS.getSupportedCodes()
* A GUI test that adds a built-in valid griset as a valid TRCS, and after the submission, the CRS class can find it

Reply:
Tests Added.